crates/runner/src/memory/mod.rs

```diff
@@ -87,20 +88,28 @@ impl Memory {
         // Get opcode from first M31
         let opcode = first_qm31.to_m31_array()[0].0;
 
-        // Determine instruction size - if opcode is invalid, just return the first QM31's M31s
-        // The VM will handle the invalid opcode error
-        let size_in_m31s = match Instruction::size_in_m31s_for_opcode(opcode) {
-            Some(size) => size,
-            None => {
-                // Invalid opcode - return just the first QM31's M31 values
-                // The VM will validate and return the proper error
-                return Ok(first_qm31.to_m31_array().to_vec());
+        // Determine instruction size using const lookup table
+        // If opcode is invalid, just return the first QM31's M31s
+        let size_in_m31s = if opcode < OPCODE_SIZE_TABLE.len() as u32 {
+            match OPCODE_SIZE_TABLE[opcode as usize] {
+                Some(size) => size,
+                None => {
+                    // Invalid opcode - return just the first QM31's M31 values
+                    // The VM will validate and return the proper error
+                    let mut result = SmallVec::new();
+                    result.extend_from_slice(&first_qm31.to_m31_array());
```

Comment:
line 89 we do `        let opcode = first_qm31.to_m31_array()[0].0;` we could cache the intermediate result and have

`let qm31_decomposed = first_qm31.to_m31_array`
`let opcode = qm31_decomposed[0].0`
...
`result.extend_from_slice(&qm31_decomposed);`